src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/ParallelDataCollectionEventsHandler.cs

```diff
@@ -0,0 +1,67 @@
+{
+    using System.Collections.Generic;
+    using System.Collections.ObjectModel;
+    using System.Linq;
+    using Microsoft.VisualStudio.TestPlatform.CommunicationUtilities;
+    using Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.Interfaces;
+    using Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.Parallel;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.Client;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.Engine;
+    using Microsoft.VisualStudio.TestPlatform.Utilities;
+
+
+    internal class ParallelDataCollectionEventsHandler : ParallelRunEventsHandler
+    {
+        private readonly ParallelRunDataAggregator runDataAggregator;
+
+        public ParallelDataCollectionEventsHandler(IProxyExecutionManager proxyExecutionManager, 
+            ITestRunEventsHandler actualRunEventsHandler, 
+            IParallelProxyExecutionManager parallelProxyExecutionManager, 
+            ParallelRunDataAggregator runDataAggregator) : 
+            this(proxyExecutionManager, actualRunEventsHandler, parallelProxyExecutionManager, runDataAggregator, JsonDataSerializer.Instance)
+        {
+        }
+
+        internal ParallelDataCollectionEventsHandler(IProxyExecutionManager proxyExecutionManager,
+            ITestRunEventsHandler actualRunEventsHandler,
+            IParallelProxyExecutionManager parallelProxyExecutionManager,
+            ParallelRunDataAggregator runDataAggregator,
+            IDataSerializer dataSerializer) : 
+            base(proxyExecutionManager, actualRunEventsHandler, parallelProxyExecutionManager, runDataAggregator, dataSerializer)
+        {
+            this.runDataAggregator = runDataAggregator;
+        }
+
+        /// <summary>
+        /// Handles the Run Complete event from a parallel proxy manager
+        /// </summary>
+        public override void HandleTestRunComplete(
+            TestRunCompleteEventArgs testRunCompleteArgs,
+            TestRunChangedEventArgs lastChunkArgs,
+            ICollection<AttachmentSet> runContextAttachments,
+            ICollection<string> executorUris)
+        {
+            var parallelRunComplete = HandleSingleTestRunComplete(testRunCompleteArgs, lastChunkArgs, runContextAttachments, executorUris);
+            
+            if (parallelRunComplete)
+            {
+                var coverageHandler = new CodeCoverageDataAttachmentsHandler();
+                var attachments = coverageHandler.HandleDataCollectionAttachmentSets(runDataAggregator.RunContextAttachments);
+
+                var completedArgs = new TestRunCompleteEventArgs(this.runDataAggregator.GetAggregatedRunStats(),
+                    this.runDataAggregator.IsCanceled,
+                    this.runDataAggregator.IsAborted,
+                    this.runDataAggregator.GetAggregatedException(),
+                    new Collection<AttachmentSet>(attachments.ToArray()),
```

Comment:
1. How are you handling attachments that are not generated by Data Collectors.
E.g. Attachment generated by TRX logger.
2. After combining attachments, the combined attachments should be replaced by singe attachment for a given DataCollector. 
3. How are we enforcing that combined attachments are removed and new attachment is passed back from the IDataCollectorAttachments interface.